File: python/src/agent_tools/bibliography.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from typing import Any, Iterable
from rapidfuzz import fuzz, process

CslEntry = dict[str, Any]

def normalize_search_text(value: Any) -> str:
    text = unicodedata.normalize("NFKD", str(value or ""))
    text = text.encode("ascii", "ignore").decode("ascii").lower()
    text = re.sub(r"[^a-z0-9]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()

def compact_search_text(value: Any) -> str:
    return re.sub(r"[^a-z0-9]+", "", normalize_search_text(value))

def _tokens(value: Any) -> list[str]:
    normalized = normalize_search_text(value)
    return normalized.split() if normalized else []

def _token_matches(token: str, field_tokens: list[str]) -> bool:
    return any(candidate == token or candidate.startswith(token) for candidate in field_tokens)

def _plausible_token_matches(token: str, field_tokens: list[str]) -> bool:
    return any(candidate == token or candidate.startswith(token) or fuzz.ratio(token, candidate) >= 82 for candidate in field_tokens)

@dataclass(frozen=True, slots=True)
class SearchHit:
    entry: CslEntry
    score: float
    match_basis: tuple[str, ...]

class BibliographyIndex:
    """Pure in-memory bibliography ranker; source/file policy lives elsewhere."""
    def __init__(self, entries: Iterable[CslEntry]):
        self.entries = list(entries)
        self._search_corpus: list[str] = []
        self._search_records: list[dict[str, Any]] = []
        for entry in self.entries:
            year = self._get_year(entry); authors = self._get_authors(entry); title = str(entry.get("title") or ""); key = str(entry.get("id") or entry.get("citation-key") or "")
            full = " ".join([year, authors, title, key]); self._search_corpus.append(full)
            self._search_records.append({"entry":entry,"key":key,"key_norm":normalize_search_text(key),"key_compact":compact_search_text(key),"key_tokens":_tokens(key),"authors":authors,"authors_norm":normalize_search_text(authors),"authors_tokens":_tokens(authors),"title":title,"title_norm":normalize_search_text(title),"title_tokens":_tokens(title),"year":year,"year_norm":normalize_search_text(year),"year_tokens":_tokens(year),"full":full,"full_norm":normalize_search_text(full),"full_tokens":_tokens(full)})
    @classmethod
    def from_entries(cls, entries: Iterable[CslEntry]) -> "BibliographyIndex": return cls(entries)
    def __len__(self) -> int: return len(self.entries)
    def has_plausible_match(self, query: str) -> bool:
        if not query.strip(): return bool(self.entries)
        info = self._query_info(query)
        return any(self._is_key_candidate(info, record) or (info["tokens"] and all(_plausible_token_matches(token, record["full_tokens"]) for token in info["tokens"])) for record in self._search_records)
    def search(self, query: str, limit: int = 20) -> list[CslEntry]: return [hit.entry for hit in self.search_hits(query, limit)]
    def search_hits(self, query: str, limit: int = 20) -> list[SearchHit]:
        if not self.entries or limit <= 0: return []
        if not query: return [SearchHit(entry=e, score=0.0, match_basis=()) for e in self.entries[:limit]]
        q = self._query_info(query); candidates = self._candidate_indexes(q, limit)
        ranked = sorted(((idx,self._search_records[idx],self._score_record(q,self._search_records[idx])) for idx in candidates), key=lambda row:(-row[2],row[0]))
        return [SearchHit(entry=record["entry"], score=float(score), match_basis=self._match_basis(q,record)) for _,record,score in ranked[:limit]]
    @staticmethod
    def _get_authors(entry):
        authors=entry.get("author")
        return " ".join(str(a.get("family","")) for a in authors if isinstance(a,dict)) if isinstance(authors,list) else ""
    @staticmethod
    def _get_year(entry):
        try:
            issued=entry.get("issued"); parts=issued.get("date-parts") if issued else None; first=parts[0] if parts else None
            return str(first[0]) if isinstance(first,list) and first else ""
        except Exception: return ""
    @staticmethod
    def _query_info(query): return {"raw":str(query or ""),"norm":normalize_search_text(query),"compact":compact_search_text(query),"tokens":_tokens(query)}
    def _candidate_indexes(self,q,limit):
        candidate_limit=min(len(self.entries),max(limit*10,200)); candidates={idx for _,_,idx in process.extract(q["raw"],self._search_corpus,limit=candidate_limit,scorer=fuzz.partial_ratio)}
        for idx,record in enumerate(self._search_records):
            if self._is_key_candidate(q,record) or self._has_full_token_coverage(q,record): candidates.add(idx)
        if len(candidates)<limit: candidates.update(range(min(len(self.entries),limit)))
        return candidates
    @staticmethod
    def _is_key_candidate(q,r):
        qc=q["compact"]; kc=r["key_compact"]; return bool(qc and kc and (qc==kc or kc.startswith(qc) or qc in kc))
    @staticmethod
    def _has_full_token_coverage(q,r): return bool(q["tokens"]) and all(_token_matches(t,r["full_tokens"]) for t in q["tokens"])
    @staticmethod
    def _score_record(q,r):
        qn=q["norm"]; qc=q["compact"]; qt=q["tokens"]
        score=max(fuzz.WRatio(qn,r["full_norm"])*.55,fuzz.partial_ratio(qn,r["full_norm"])*.45,fuzz.token_set_ratio(qn,r["full_norm"])*.55,fuzz.WRatio(qn,r["key_norm"])*.90,fuzz.WRatio(qn,r["authors_norm"])*.75,fuzz.WRatio(qn,r["title_norm"])*.70)
        if qc and r["key_compact"]:
            score=max(score,fuzz.partial_ratio(qc,r["key_compact"])*.95,fuzz.WRatio(qc,r["key_compact"])*.90)
            if qc==r["key_compact"]: score+=50
            elif len(qc)>=3 and r["key_compact"].startswith(qc): score+=25
        if not qt: return float(score)
        score += 8*sum(1 for t in qt if _token_matches(t,r["full_tokens"]))/len(qt)
        if any(_token_matches(t,r["authors_tokens"]) for t in qt) and any(_token_matches(t,r["title_tokens"]) or _token_matches(t,r["key_tokens"]) for t in qt): score+=12
        if any(_token_matches(t,r["year_tokens"]) for t in qt): score+=8
        return float(score)
    @staticmethod
    def _match_basis(q,r):
        basis=set(); qc=q["compact"]; kc=r["key_compact"]
        if qc and kc and (qc==kc or (len(qc)>=3 and kc.startswith(qc))): basis.add("key")
        if any(_token_matches(t,r["authors_tokens"]) for t in q["tokens"]): basis.add("author")
        if any(_token_matches(t,r["title_tokens"]) for t in q["tokens"]): basis.add("title")
        if any(_token_matches(t,r["year_tokens"]) for t in q["tokens"]): basis.add("year")
        return tuple(sorted(basis))

File: typescript/src/bibliography.ts
import fuzzysort from 'fuzzysort';

export interface CslEntry {
  id?: string | number;
  type?: string;
  title?: string;
  author?: { family?: string; given?: string }[];
  issued?: { 'date-parts'?: (number | string)[][] };
  'citation-key'?: string | number;
  [key: string]: any;
}

export interface SearchHit {
  entry: CslEntry;
  score: number;
  matchBasis: string[];
}

interface SearchRecord {
  index: number;
  obj: CslEntry;
  key: string;
  keyNorm: string;
  keyCompact: string;
  keyTokens: string[];
  authors: string;
  authorsNorm: string;
  authorsTokens: string[];
  title: string;
  titleNorm: string;
  titleTokens: string[];
  year: string;
  yearNorm: string;
  yearTokens: string[];
  full: string;
  fullNorm: string;
  fullTokens: string[];
}

interface QueryInfo {
  raw: string;
  norm: string;
  compact: string;
  tokens: string[];
}

/**
 * Pure in-memory bibliography ranker.
 *
 * Filesystem/source-role policy deliberately lives in ReferenceCatalog. This
 * class only indexes an already chosen set of immutable entries and preserves
 * the ranking semantics historically used by agent-tools/bibq-w.
 */
export class BibliographyIndex {
  public readonly entries: CslEntry[];
  private readonly searchIndex: SearchRecord[];

  private constructor(entries: readonly CslEntry[]) {
    this.entries = entries.slice();
    this.searchIndex = this.entries.map((entry, index) => this.buildSearchRecord(entry, index));
  }

  static fromEntries(entries: readonly CslEntry[]): BibliographyIndex {
    return new BibliographyIndex(entries);
  }

  get length(): number {
    return this.entries.length;
  }

  /**
   * Match evidence independent of search()'s legacy padded candidates.
   * Accepts normalized key containment or coverage of every query token using
   * the same prefix/typo matcher as ranking. Blank queries match nonempty indexes.
   * This is a widening heuristic, not a filter or a guarantee of relevance.
   */
  hasPlausibleMatch(query: string): boolean {
    if (!query.trim()) return this.entries.length > 0;
    const info = this.getQueryInfo(query);
    return this.searchIndex.some(record => this.isKeyCandidate(info, record) ||
      (info.tokens.length > 0 && info.tokens.every(token => tokenMatches(token, record.fullTokens))));
  }

  search(query: string, limit: number = 20): CslEntry[] {
    return this.searchHits(query, limit).map(hit => hit.entry);
  }

  /**
   * Ranked results with raw score and non-exclusive field evidence.
   * Scores are implementation details and are not promised to be comparable
   * across language bindings. Ordering is exactly the same as search().
   */
  searchHits(query: string, limit: number = 20): SearchHit[] {
    if (this.entries.length === 0 || limit <= 0) return [];
    if (!query) {
      return this.entries.slice(0, limit).map(entry => ({ entry, score: 0, matchBasis: [] }));
    }

    const queryInfo = this.getQueryInfo(query);
    const candidateIndexes = this.candidateIndexes(queryInfo, limit);
    return Array.from(candidateIndexes)
      .map(index => {
        const record = this.searchIndex[index];
        return {
          record,
          index,
          score: this.scoreRecord(queryInfo, record),
          matchBasis: this.matchBasis(queryInfo, record),
        };
      })
      .sort((a, b) => b.score - a.score || a.index - b.index)
      .slice(0, limit)
      .map(result => ({ entry: result.record.obj, score: result.score, matchBasis: result.matchBasis }));
  }

  private buildSearchRecord(e: CslEntry, index: number): SearchRecord {
    const year = this.getYear(e);
    const authors = this.getAuthors(e);
    const title = String(e.title || '');
    const key = String(e.id || e['citation-key'] || '');
    const full = `${year} ${authors} ${title} ${key}`;
    return {
      index, obj: e, key,
      keyNorm: normalize(key), keyCompact: compact(key), keyTokens: tokens(key),
      authors, authorsNorm: normalize(authors), authorsTokens: tokens(authors),
      title, titleNorm: normalize(title), titleTokens: tokens(title),
      year, yearNorm: normalize(year), yearTokens: tokens(year),
      full, fullNorm: normalize(full), fullTokens: tokens(full),
    };
  }

  private getAuthors(e: CslEntry): string {
    if (!Array.isArray(e.author)) return '';
    return e.author.map(author => author?.family || '').join(' ');
  }

  private getYear(e: CslEntry): string {
    try {
      const parts = e.issued?.['date-parts'];
      if (parts && Array.isArray(parts) && parts.length > 0) {
        const first = parts[0];
        if (Array.isArray(first) && first.length > 0) return String(first[0]);
      }
    } catch {
      // Ignore malformed date data; the historical ranker did the same.
    }
    return '';
  }

  private getQueryInfo(query: string): QueryInfo {
    return { raw: String(query || ''), norm: normalize(query), compact: compact(query), tokens: tokens(query) };
  }

  private candidateIndexes(query: QueryInfo, limit: number): Set<number> {
    const candidateLimit = Math.min(this.entries.length, Math.max(limit * 10, 200));
    const candidates = new Set<number>();
    const fuzzyResults = fuzzysort.go(query.raw, this.searchIndex, { key: 'full', limit: candidateLimit, threshold: -Infinity });
    fuzzyResults.forEach(result => candidates.add(result.obj.index));
    this.searchIndex.forEach(record => {
      if (this.isKeyCandidate(query, record) || this.hasFullTokenCoverage(query, record)) candidates.add(record.index);
    });
    // Historical behavior: pad with initial entries. ReferenceCatalog uses
    // hasPlausibleMatch() separately when it needs relevance evidence.
    if (candidates.size < limit) {
      for (let index = 0; index < Math.min(this.entries.length, limit); index += 1) candidates.add(index);
    }
    return candidates;
  }

  private isKeyCandidate(query: QueryInfo, record: SearchRecord): boolean {
    return Boolean(query.compact && record.keyCompact && (
      query.compact === record.keyCompact ||
      record.keyCompact.startsWith(query.compact) ||
      record.keyCompact.includes(query.compact)
    ));
  }

  private hasFullTokenCoverage(query: QueryInfo, record: SearchRecord): boolean {
    return query.tokens.length > 0 && query.tokens.every(token => strictTokenMatches(token, record.fullTokens));
  }

  private scoreRecord(query: QueryInfo, record: SearchRecord): number {
    let score = Math.max(
      similarity(query.norm, record.fullNorm) * 0.55,
      containsSimilarity(query.norm, record.fullNorm) * 0.45,
      tokenSetSimilarity(query.norm, record.fullNorm) * 0.55,
      similarity(query.norm, record.keyNorm) * 0.90,
      similarity(query.norm, record.authorsNorm) * 0.75,
      similarity(query.norm, record.titleNorm) * 0.70,
    );
    if (query.compact && record.keyCompact) {
      score = Math.max(score,
        partialSimilarity(query.compact, record.keyCompact) * 0.95,
        similarity(query.compact, record.keyCompact) * 0.90,
      );
      if (query.compact === record.keyCompact) score += 50;
      else if (query.compact.length >= 3 && record.keyCompact.startsWith(query.compact)) score += 25;
    }
    if (query.tokens.length === 0) return score;
    const covered = query.tokens.filter(token => tokenMatches(token, record.fullTokens)).length;
    score += 8 * covered / query.tokens.length;
    const hasAuthorToken = query.tokens.some(token => tokenMatches(token, record.authorsTokens));
    const hasTitleOrKeyToken = query.tokens.some(token => tokenMatches(token, record.titleTokens) || tokenMatches(token, record.keyTokens));
    if (hasAuthorToken && hasTitleOrKeyToken) score += 12;
    if (query.tokens.some(token => tokenMatches(token, record.yearTokens))) score += 8;
    return score;
  }

  private matchBasis(query: QueryInfo, record: SearchRecord): string[] {
    const basis = new Set<string>();
    if (query.compact && record.keyCompact && (
      query.compact === record.keyCompact ||
      (query.compact.length >= 3 && record.keyCompact.startsWith(query.compact))
    )) basis.add('key');
    if (query.tokens.some(token => tokenMatches(token, record.authorsTokens))) basis.add('author');
    if (query.tokens.some(token => tokenMatches(token, record.titleTokens))) basis.add('title');
    if (query.tokens.some(token => tokenMatches(token, record.yearTokens))) basis.add('year');
    return Array.from(basis).sort();
  }
}

export function normalizeSearchText(value: unknown): string {
  return String(value || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

export function compactSearchText(value: unknown): string {
  return normalizeSearchText(value).replace(/[^a-z0-9]+/g, '');
}

function normalize(value: unknown): string { return normalizeSearchText(value); }
function compact(value: unknown): string { return compactSearchText(value); }
function tokens(value: unknown): string[] { const normalized = normalize(value); return normalized ? normalized.split(' ') : []; }
function tokenMatches(token: string, candidates: string[]): boolean {
  return candidates.some(candidate => candidate === token || candidate.startsWith(token) || similarity(token, candidate) >= 82);
}
function strictTokenMatches(token: string, candidates: string[]): boolean {
  return candidates.some(candidate => candidate === token || candidate.startsWith(token));
}
function tokenSetSimilarity(a: string, b: string): number {
  const aTokens = tokens(a); const bTokens = tokens(b);
  if (aTokens.length === 0 || bTokens.length === 0) return 0;
  const total = aTokens.reduce((sum, token) => {
    const best = bTokens.reduce((max, candidate) => Math.max(max, tokenMatches(token, [candidate]) ? 100 : similarity(token, candidate)), 0);
    return sum + best;
  }, 0);
  return total / aTokens.length;
}
function partialSimilarity(needle: string, haystack: string): number {
  if (!needle || !haystack) return 0;
  if (haystack.includes(needle)) return 100;
  if (needle.length > haystack.length) return similarity(needle, haystack);
  let best = 0;
  for (let i = 0; i <= haystack.length - needle.length; i += 1) best = Math.max(best, similarity(needle, haystack.slice(i, i + needle.length)));
  return best;
}
function containsSimilarity(needle: string, haystack: string): number { return !needle || !haystack ? 0 : haystack.includes(needle) ? 100 : 0; }
function similarity(a: string, b: string): number {
  if (!a && !b) return 100;
  if (!a || !b) return 0;
  if (a === b) return 100;
  const distance = levenshtein(a, b);
  return Math.max(0, 100 * (1 - distance / Math.max(a.length, b.length)));
}
function levenshtein(a: string, b: string): number {
  const previous = Array.from({ length: b.length + 1 }, (_, index) => index);
  const current = Array.from({ length: b.length + 1 }, () => 0);
  for (let i = 1; i <= a.length; i += 1) {
    current[0] = i;
    for (let j = 1; j <= b.length; j += 1) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      current[j] = Math.min(current[j - 1] + 1, previous[j] + 1, previous[j - 1] + cost);
    }
    for (let j = 0; j <= b.length; j += 1) previous[j] = current[j];
  }
  return previous[b.length];
}
